### PR TITLE
Add memcpy with bounds checks.

### PR DESCRIPTION
**Pull request by davidchisnall**
The memcpy implementation is not completely stupid but is almost
certainly not as good as a carefully tuned and optimised one.

Building snmalloc with FreeBSD's libc memcpy + jemalloc and with this,
each 10 times, does not show a statistically significant performance
difference at 95% confidence.  The snmalloc version has very slightly
lower median and worst-case times.  This is in no way a sensible
benchmark, but it serves as a smoke test for significant performance
regressions.

The CI self-host job now uses the checked memcpy.

This also fixes an off-by-one error in the external bounds.  This is
triggered by ninja, so we will see breakage in CI if it is reintroduced.

In debug builds, we provide a verbose error containing the address of
the allocation, the base and bounds of the allocation, and a backtrace.

The backtrace was broken by the CI cleanup moving the BACKTRACE_HEADER
macro into the SNMALLOC_ namespace.  This is also fixed.

### CODE REVIEW

**Comment by mjp41** [comment]
Sorry, could you add some test coverage.  We really need decent coverage and testing of a memcpy implementation.

It would be good to hook up some negative test coverage too.  Though, as we don't have anything like that at the moment, I would be happy to leave that to a separate PR, where we add negative test coverage for many of the security features.

**Comment by davidchisnall** [comment]
We had an integration test (the self-host build now uses the checked memcpy) but I've also added a unit test.  This checks that we catch the bounds errors and that memcpy does the right thing in fun corner cases.

File: src/override/memcpy.cc [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+  template<bool IsRead = false>
+  SNMALLOC_FAST_PATH inline void
+  check_bounds(const void* ptr, size_t len, const char* msg = "")
+  {
+    if constexpr (!IsRead || CheckReads)
+    {
+      auto& alloc = ThreadAlloc::get();
+      void* p = const_cast<void*>(ptr);
+
+      // FIXME: Overflow checking.
```
Comment (mjp41) [inline src/override/memcpy.cc:139]: Could you use `TODO` rather than `FIXME`?

File: src/backend/pagemap.h [hunk 2]
```diff
@@ -244,7 +244,14 @@ namespace snmalloc
       }
 
       //  This means external pointer on Windows will be slow.
-      if constexpr (potentially_out_of_range && !pal_supports<LazyCommit, PAL>)
+      // FIXME: With SNMALLOC_CHECK_CLIENT we should map that region read-only,
```
Comment (mjp41) [inline src/backend/pagemap.h:247]: Could you use `TODO` rather `FIXME`.

File: src/override/memcpy.cc [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+      if (unlikely(
+            pointer_offset(ptr, len) > alloc.external_pointer<OnePastEnd>(p)))
```
Comment (mjp41) [inline src/override/memcpy.cc:141]: Isn't the check you actually want:
```
pointer_diff(ptr, alloc.external_pointer<OnePastEnd>(p)) > len
```
then it works for overflow.

**Review by mjp41 — APPROVED** [review]
LGTM. Couple of minor bits

File: src/override/memcpy.cc (hunk 1 above)
Comment (davidchisnall) [reply src/override/memcpy.cc:139]: I'd prefer to keep FIXME for things that are really important.  The lack of overflow checking is actually a security vulnerability here.  To me, TODO implies 'we should probably do this', FIXME means 'this is really a bug and we should fix it'.

File: src/override/memcpy.cc (hunk 3 above)
Comment (davidchisnall) [reply src/override/memcpy.cc:141]: Yes, if you replace `>` with `<`, that's better.  Thanks!

File: src/override/memcpy.cc [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+    {
+      char data[Size];
+    };
+    auto* d = static_cast<Block*>(dst);
+    auto* s = static_cast<const Block*>(src);
+    *d = *s;
+#endif
+  }
+
+  SNMALLOC_SLOW_PATH SNMALLOC_UNUSED_FUNCTION void crashWithMessage
```
[the remark this answers is not in the document]
Comment (davidchisnall) [reply src/override/memcpy.cc:102]: Could be if we have other consumers of it.  If we want to add more things with bounds checks, it definitely should be.

File: src/override/memcpy.cc (hunk 3 above)
Comment (mjp41) [reply src/override/memcpy.cc]: I was to busy focussing on getting the `pointer_diff` argument order correct to get that right ;-)

File: src/override/memcpy.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (mjp41) [reply src/override/memcpy.cc:102]: Well, there are a bunch of things that do checks

* freeing the start of object
* corruption of free list

So, giving those nicer error message would be good.  But clearly a different PR.  Let's leave that refactoring to then.

[the remark this answers is not in the document]
Comment (davidchisnall) [reply src/override/memcpy.cc:102]: Sounds good.  This function is quite specialised because it's trying to keep everything related to preparing the message in the slow-path function, so that we can just bail out and do all of the complicated formatting things in a separate function.  It would probably need to be refactored to be a generic that took a lambda that did the formatting if we wanted something more general, ideally with the fast-path and slow-path bits both being generic.  I'll have a think about what that should look like.
